### PR TITLE
Add kn-go-harbor-webhook function

### LINKED ISSUE #914: Create a custom webhook function for Harbor
**Is your feature request related to a problem? Please describe.**
Since Harbor is providing a [webhook feature](https://goharbor.io/docs/2.2.0/working-with-projects/project-configuration/configure-webhooks/) for each created project/repository, it'd be great to have a custom webhook-function available which provides us custom Harbor specific events like e.g. `com.vmware.harbor.artifact_pushed.v0`.

**Describe the solution you'd like**
By following @lamw's provided way to create a genereic webhook function [HERE](https://williamlam.com/2021/09/custom-webhook-function-to-publish-events-into-vmware-event-broker-appliance-veba.html), I was able to set up the webhook config in Harbor properly in order to send events to it.

Figure I:
image

When an event is sent based on an action such as e.g. `docker push`, the following very usable output appears in the webhook-deployment `user-container`.

`k -n vmware-functions logs kn-ps-webhook-00002-deployment-8f8664b49-6mhnl user-container`

```
user-container 06/22/2022 09:01:43 - Creating CloudEvent ...
user-container
user-container 06/22/2022 09:01:43 - [Headers]:
user-container
user-container Name                           Value
user-container ----                           -----
user-container X-Forwarded-For                {10.197.96.51, 10.10.0.1, 10.10.0.30
user-container Content-Type                   {application/json
user-container Forwarded                      {for=10.197.96.51;proto=http, for=10.10.0.1
user-container Host                           {kn-ps-webhook.vmware-functions.veba-dev.jarvis.tanzu
user-container User-Agent                     {Go-http-client/1.1
user-container X-Forwarded-Proto              {http
user-container X-Request-Id                   {94817867-aced-450b-a4d4-e046389e5475
user-container X-Envoy-Internal               {true
user-container Accept-Encoding                {gzip
user-container K-Proxy-Request                {activator
user-container Content-Length                 {448
user-container X-Request-Start                {t=1655888503.111
user-container
user-container
user-container 06/22/2022 09:01:43 - [Body]:
user-container
user-container @{type=PUSH_ARTIFACT; occur_at=1655888502; operator=admin; event_data=}
user-container
user-container 06/22/2022 09:01:43 - [CloudEvent]:
user-container
user-container DataContentType : application/json
user-container Data            : {123, 34, 116, 121…
user-container Id              : c025833f-7c0e-44b1-9926-52fa92c45443
user-container DataSchema      :
user-container Source          : 10.197.96.51
user-container SpecVersion     : V1_0
user-container Subject         :
user-container Time            : 06/22/2022 09:01:43
user-container Type            : com.vmware.generic.webhook.v0
user-container
user-container
user-container
user-container 06/22/2022 09:01:43 - Sending CloudEvent to VMware Event Broker ...
user-container
user-container 06/22/2022 09:01:43 - Successfully sent CloudEvent ...
user-container
```
Even better is the output in Sockeye 😃 

Figure II:

image

Let's create a function which gives us the possibility to process those custom events.

### PR DESCRIPTION
**Pull request by embano1**
Closes: #914
Signed-off-by: Michael Gasch <mgasch@vmware.com>

## Summary

Please describe your changes and what you plan to introduce as part of this pull request submission. If it fixes a bug or resolves a feature request, be sure to link to the issue numbers following the template below.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [x] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [x] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [ ] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [x] Please include any relevant screenshots and/or output as part of your testing
- [x] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [ ] Other (please describe):

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

Closes #914

## Testing Verification

Verified in real environment with @rguske .

Added unit tests:

```console
go test -v -race -count 1 ./... -json -cover | tparse -all

+--------+---------+---------------------------------------------------------+----------------------+
| STATUS | ELAPSED |                          TEST                           |       PACKAGE        |
+--------+---------+---------------------------------------------------------+----------------------+
| PASS   |    1.00 | Test_run                                                | kn-go-harbor-webhook |
| PASS   |    0.00 | Test_eventhandler                                       | kn-go-harbor-webhook |
| PASS   |    0.00 | Test_eventhandler/successfully_sends_event_(no_auth)    | kn-go-harbor-webhook |
| PASS   |    0.00 | Test_eventhandler/fails_with_401_not_authorized         | kn-go-harbor-webhook |
| PASS   |    0.00 | Test_eventhandler/successfully_sends_event_(basic_auth) | kn-go-harbor-webhook |
+--------+---------+---------------------------------------------------------+----------------------+

+--------+---------+---------------------------------------------------------------------------------------------------+-------+------+------+------+
| STATUS | ELAPSED |                                              PACKAGE                                              | COVER | PASS | FAIL | SKIP |
+--------+---------+---------------------------------------------------------------------------------------------------+-------+------+------+------+
| PASS   | 1.36s   | github.com/vmware-samples/vcenter-event-broker-appliance/examples/knative/go/kn-go-harbor-webhook | 61.4% |    5 |    0 |    0 |
+--------+---------+---------------------------------------------------------------------------------------------------+-------+------+------+------+
```

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

### CODE REVIEW

File: examples/knative/go/kn-go-harbor-webhook/function.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    app: veba-ui
+spec:
+  template:
+    metadata:
+      annotations:
+        autoscaling.knative.dev/maxScale: "1"
+        autoscaling.knative.dev/minScale: "1"
+    spec:
+      containers:
+        - image: ko://github.com/vmware-samples/vcenter-event-broker-appliance/examples/knative/go/kn-go-harbor-webhook
```
Comment (embano1) [inline examples/knative/go/kn-go-harbor-webhook/function.yaml:15]: fix

File: examples/knative/go/kn-go-harbor-webhook/README.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
+# kn-go-nsx-tag-sync
```
Comment (embano1) [inline examples/knative/go/kn-go-harbor-webhook/README.md:1]: fix

File: examples/knative/go/kn-go-harbor-webhook/sinkbinding.yaml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  subject:
+    apiVersion: serving.knative.dev/v1
+    kind: Service
+    name: kn-go-harbor-webhook
+  sink:
+    ref:
+      apiVersion: eventing.knative.dev/v1
+      kind: Broker
+      name: example-broker
+      namespace: default
```
Comment (embano1) [inline examples/knative/go/kn-go-harbor-webhook/sinkbinding.yaml:15]: fix

File: examples/knative/go/kn-go-harbor-webhook/README.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+```bash
+# undeploy function
+kubectl delete -f function.yaml -n vmware-functions
+
+# undeploy sinkbinding
+kubectl delete -f sinkbinding.yaml -n vmware-functions
+
+# delete secret
+kubectl delete secret vsphere-credentials -n vmware-functions
```
Comment (rguske) [inline examples/knative/go/kn-go-harbor-webhook/README.md:218]: Secret will be `webhook-auth`.

File: examples/knative/go/kn-go-harbor-webhook/README.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+Deploy the function to the VMware Event Broker Appliance (VEBA):
+
+```bash
+kubectl create -f function.yaml -n vmware-functions
+```
+
+For testing purposes, the [Knative manifest](function.yaml) contains the
+following annotations, which will ensure the Knative Service Pod will always run
+**exactly** one instance for debugging purposes. Functions deployed through
+through the VMware Event Broker Appliance UI defaults to scale to 0, which means
```
Comment (rguske) [inline examples/knative/go/kn-go-harbor-webhook/README.md:179]: NIT: 'through' is doubled

**Review by rguske — CHANGES_REQUESTED** [review]
Please see comments in README.md.

File: examples/knative/go/kn-go-harbor-webhook/README.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+--from-literal=password='replaceme'
+--namespace vmware-functions
+
+# update label for secret to show up in VEBA UI
+kubectl -n vmware-functions label secret webhook-auth app=veba-ui
+```
+
+## Update Environment Settings
+
+The `function.yaml` comes with sane defaults, incl. basic auth for the HTTP
```
[the remark this answers is not in the document]
Comment (embano1) [reply examples/knative/go/kn-go-harbor-webhook/README.md:142]: Nope, "sane" is what I meant.

**Comment by embano1** [comment]
@rguske comments addressed, thx

File: examples/knative/go/kn-go-harbor-webhook/README.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
+# kn-go-harbor-webhook
+
+Example Knative Go function for receiving [Project Harbor
+webhook](https://goharbor.io/docs/1.10/working-with-projects/project-configuration/configure-webhooks/)
```
Comment (rguske) [inline examples/knative/go/kn-go-harbor-webhook/README.md:4]: Link is referencing to an old (`v1.10`) documentation page. Event naming has changed for Harbor version `v2.5x`. Please use https://goharbor.io/docs/2.5.0/working-with-projects/project-configuration/configure-webhooks/

Comment (embano1) [reply examples/knative/go/kn-go-harbor-webhook/README.md:4]: Good catch! Using "latest" now to always ref latest docs.

**Review by rguske — APPROVED** [review]
LGTM

Great work!

**Pull request merged**